RavenDB-18164 - Query against a map-reduce index should get all the results

grisha-kotler opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-18164/Query-against-a-map-reduce-index-should-get-all-the-results
https://issues.hibernatingrhinos.com/issue/RavenDB-19975/Sharding-Querying-Filter-doesnt-work

### Additional description

- Support for `filter` in map-reduce queries.
- Fix the `ScannedResults` in sharded queries.
- Support for multiple `group by` fields in an auto map-reduce query.
- Support for `limit` and `order by` in map-reduce queries.

### Type of change

- New feature

### How risky is the change?

- Low 

### Backward compatibility

- Non breaking change

### Documentation update

- This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.

### Testing by Contributor

- Tests have been added that prove the fix is effective or that the feature works

grisha-kotler committed:
RavenDB-18164 - add support for filter in sharded queries

grisha-kotler committed:
RavenDB-18164 - fill the ScannedResults

grisha-kotler committed:
RavenDB-18164 - fix queries with limit

grisha-kotler committed:
RavenDB-18164 - add support for order by queries with limit

ayende commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -213,24 +218,81 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+                            throw new InvalidOperationException($"Index '{Query.Metadata.IndexName}' is not a map-reduce index");
+
+                        var compiled = ((StaticIndexInformationHolder)index).Compiled;
+                        groupByFields = compiled.GroupByFields.Select(x => x.Name).ToList();
```
This includes JavaScript indexes?

grisha-kotler replied:
yes. added tests for that.

ayende commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -213,24 +218,81 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+                    // we need to get all results if the order by isn't done on the group by fields
+
+                    var orderByFields = Query.Metadata.OrderBy.Select(x => x.Name.Value).ToList();
+                    rewriteForLimitWithOrderByInMapReduce = groupByFields.Count != orderByFields.Count ||
+                                                            groupByFields.Intersect(orderByFields).Count() != orderByFields.Count;
```
I'm not following this logic

grisha-kotler replied:
I want to compare the `group by` fields with the `order by` fields.
If `order by` fields contains all the group by fields than we can send the `limit`.

ayende commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -213,24 +218,81 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+                    // we have a limit and order by fields
+                    // we need to get all results if the order by isn't done on the group by fields
+
+                    var orderByFields = Query.Metadata.OrderBy.Select(x => x.Name.Value).ToList();
```
All of those `ToList()` can be expensive, can we move that to the query metadata so it only happens once?

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -213,24 +218,81 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+
+                if (IsAutoMapReduceQuery)
+                {
+                    groupByFields = Query.Metadata.GroupBy.Select(x => x.Name.Value).ToList();
```
it probably means that we can cache it?

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -252,13 +314,48 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+        else if (setOrderByFieldsForMapReduce != null)
+        {
+            clone.OrderBy = new List<(QueryExpression Expression, OrderByFieldType FieldType, bool Ascending)>();
+            Query.Metadata.OrderBy = new OrderByField[setOrderByFieldsForMapReduce.Count];
```
isn't this done on a cached instance that can be shared between queries?

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -252,13 +314,48 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+            {
+                var field = setOrderByFieldsForMapReduce[i];
+                clone.OrderBy.Add((new FieldExpression(new List<StringSegment> { field }), OrderByFieldType.Implicit, Ascending: true));
+                Query.Metadata.OrderBy[i] = new OrderByField(new QueryFieldName(field, isQuoted: false), OrderByFieldType.Implicit, ascending: true);
```
isn't this done on a cached instance that can be shared between queries?

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Documents/Sharding/Operations/Queries/AbstractShardedQueryOperation.cs:
```diff
@@ -69,6 +69,12 @@ protected static void CombineSingleShardResultProperties(QueryResult<List<TResul
+        if (singleShardResult.ScannedResults != null)
+        {
+            combinedResult.ScannedResults ??= 0;
+            combinedResult.ScannedResults += singleShardResult.ScannedResults;
```
where filter is happening actually? In the orchestrator only? then it should be 0 always and the final result should get the value from orchestrator only?

grisha-kotler replied:
map indexes - on the shards
map-reduce - on the orchestrator (`singleShardResult.ScannedResults` will be null)

arekpalinski commented on test/FastTests/Sharding/Queries/BasicShardedMapReduceQueryTests.cs:
```diff
@@ -0,0 +1,923 @@
+
+            public class CompoundResult
+            {
+                public ExpandoObject Name;
```
Please check the warning here

grisha-kotler replied:
done

arekpalinski commented on test/FastTests/Sharding/Queries/BasicShardedMapReduceQueryTests.cs:
```diff
@@ -0,0 +1,923 @@
+            public class CompoundResult
+            {
+                public ExpandoObject Name;
+                public int Sum;
```
And same warning here

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -213,24 +218,81 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+                }
+                else
+                {
+                    Debug.Assert(IsMapReduceIndex);
```
Please provide message as 2nd parameter so we'll know immediately what assertion failed

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Sharding/Queries/ShardedQueryProcessor.cs:
```diff
@@ -208,6 +211,44 @@ public static async ValueTask HandleMissingTimeSeriesIncludesAsync(TransactionOp
+
+        using (scope?.For(nameof(QueryTimingsScope.Names.Filter)))
+        {
+            var key = new CollectionQueryEnumerable.FilterKey(Query.Metadata);
+            using (RequestHandler.DatabaseContext.Indexes.ScriptRunnerCache.GetScriptRunner(key, readOnly: true, patchRun: out var filterScriptRun))
+            {
+                var currentResults = result.Results;
+                result.Results = new List<BlittableJsonReaderObject>();
+
+                result.ScannedResults++;
+
+                foreach (var doc in currentResults)
+                {
+                    object self = filterScriptRun.Translate(Context, doc);
+                    using (var filterResult = filterScriptRun.Run(Context, null, "execute", new[] { self, Query.QueryParameters }, scope))
+                    {
+                        if (filterResult.BooleanValue == true)
+                        {
+                            result.Results.Add(doc);
+                        }
+                        else
+                        {
+                            result.SkippedResults++;
+                        }
+                    }
+                }
+            }
```
We have `QueryFilter` class which is used by Lucene / Corax read operations. The idea was to have the filter implementation handled in single place. 

Can we use this class and it's `Apply()` method (maybe create a new, dedicated one) instead of repeating the same code here?

arekpalinski replied:
Also I think we should use `FilterResult` enum

grisha-kotler replied:
done

arekpalinski commented on test/FastTests/Issues/RavenDB-17636.cs:
```diff
@@ -23,7 +24,7 @@ public FilterTests(ITestOutputHelper output) : base(output)
 
     [RavenTheory(RavenTestCategory.Querying)]
-    [RavenData(SearchEngineMode = RavenSearchEngineMode.All)]
+    [RavenData(SearchEngineMode = RavenSearchEngineMode.All, DatabaseMode = RavenDatabaseMode.All)]
```
Can you switch tests from `FilterTests.cs.` to be run in sharded mode as well?

grisha-kotler replied:
already done

arekpalinski commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -252,13 +314,48 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+            && rewriteForProjectionFromMapReduceIndex == false 
+            && rewriteForProjectionFromAutoMapReduceQuery == false
+            && rewriteForFilterForMapReduceIndexOrAutoMapReduceQuery == false
+            && rewriteForLimitWithOrderByInMapReduce == false
```
Can we use some enum with flags instead of those variables?

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Sharding/Queries/ShardedQueryProcessorBase.cs:
```diff
@@ -46,12 +46,13 @@ protected void ApplyPaging(ref TCombinedResult result, QueryTimingsScope scope)
+                long index = Math.Min(limit, int.MaxValue);
+                long count = result.Results.Count - limit;
                 if (count > int.MaxValue)
                     count = int.MaxValue; //todo: Grisha: take a look
```
I think this TODO can be removed. Also please check:

https://issues.hibernatingrhinos.com/issue/RavenDB-19990/Allow-to-skip-over-int.Max-docs-in-query-for-Corax-but-limit-projection-up-to-int.Max

grisha-kotler replied:
done

grisha-kotler committed:
RavenDB-18164 - refactoring

grisha-kotler committed:
RavenDB-18164 - added ShardedQueryFilter

grisha-kotler committed:
RavenDB-18164 - support JavaScript map-reduce indexes

grisha-kotler committed:
RavenDB-18164 - enable tests for sharding

grisha-kotler committed:
RavenDB-18164 - don't use the OrderBy fields from the cached query

grisha-kotler committed:
RavenDB-18164 - use flags

grisha-kotler committed:
RavenDB-18164 - remove comment

grisha-kotler committed:
RavenDB-18164 - fixing null reduce keys handling
in sharding we get the results from all the shards but they aren't processed in aggregation batches by the reduce key (like in the non-sharded version)
this is fixing a corner case where we got a sequence of: null, "not null", null reduce keys

grisha-kotler committed:
RavenDB-18164 - make the order of LastModified consistent across the shards

grisha-kotler committed:
RavenDB-18164 - remove tabs

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/QueryFilter.cs:
```diff
@@ -16,56 +17,57 @@ public enum FilterResult
 }
 
-public class QueryFilter : IDisposable
+public abstract class QueryFilterBase : IDisposable
```
AbstractQueryFilter

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/QueryFilter.cs:
```diff
@@ -75,3 +77,79 @@ public void Dispose()
     }
 }
+
+public class QueryFilter : QueryFilterBase
```
don't keep everything in the same file, move it

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Documents/Indexes/Persistence/QueryFilter.cs:
```diff
@@ -75,3 +77,79 @@ public void Dispose()
+    }
+}
+
+public class ShardedQueryFilter : QueryFilterBase
```
this is not a proper place for that, we do have a Sharding namespace available

grisha-kotler replied:
done

arekpalinski commented:
@grisha-kotler please check failing test

arekpalinski commented on src/Raven.Server/Documents/Indexes/Persistence/QueryFilter.cs:
```diff
@@ -75,3 +77,79 @@ public void Dispose()
+    }
+}
+
+public class ShardedQueryFilter : QueryFilterBase
```
Any reason why `QueryFilter` and `ShardedQueryFilter` aren't put in separate files? It's way easier to find them on github then

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -223,26 +230,74 @@ private void AssertUsingCustomSorters()
+
+            if (query.Limit != null)
+            {
+                reduceKeys = GetGroupByFields();
```
I wonder if `groupByFields` isn't more appropriate name than `reduceKeys`

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -334,6 +461,63 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+    }
+
+    [Flags]
+    private enum QueryChanges
```
Can we move it to the end of the file?

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -259,19 +314,60 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+            }
+
+            // add the missing reduce keys
+            foreach (var reduceKey in reduceKeys)
```
I think we should guarantee the order of  added group by fields. `HashSet` doesn't guarantee that

grisha-kotler replied:
done

arekpalinski commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -259,19 +314,60 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
+        if (queryChanges.HasFlag(QueryChanges.UpdateOrderByFieldsInMapReduce))
+        {
+            clone.OrderBy ??= new List<(QueryExpression Expression, OrderByFieldType FieldType, bool Ascending)>();
+            OrderByFields = new List<OrderByField>();
```
Do we need to create it every time? Can't we store it in query metadata which is cached?

grisha-kotler replied:
done

ppekrol commented on src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs:
```diff
@@ -318,14 +414,45 @@ private BlittableJsonReaderObject RewriteQueryIfNeeded(BlittableJsonReaderObject
-        else if (rewriteForProjectionFromAutoMapReduceIndex)
+        else if (queryChanges.HasFlag(QueryChanges.RewriteForProjectionFromAutoMapReduceQuery))
         {
+            var missingGroupByFieldsFromProjection = Query.Metadata.GroupBy.Select(x => x.Name.Value).ToHashSet();
```
can be cached in the metadata?

grisha-kotler replied:
done

grisha-kotler committed:
RavenDB-18164 - PR comments

grisha-kotler committed:
RavenDB-18164 - fixed failing tests

arekpalinski commented on src/Raven.Server/Documents/Indexes/Static/StaticIndexBase.cs:
```diff
@@ -200,13 +200,13 @@ public abstract class AbstractStaticIndexBase
+        private List<string> _groupByFieldNames;
 
-        public HashSet<string> GroupByFieldNames
+        public List<string> GroupByFieldNames
```
Same comment as above

arekpalinski replied:
I meant this one - https://github.com/ravendb/ravendb/pull/15936#discussion_r1138709370

arekpalinski commented on src/Raven.Server/Documents/Indexes/MapReduce/Auto/AutoMapReduceIndexDefinition.cs:
```diff
@@ -19,6 +19,16 @@ internal class AutoMapReduceIndexDefinition : AutoIndexDefinitionBaseServerSide
+            {
+                return _groupByFieldNames ??= GroupByFields.Keys.ToList();
+            }
+        }
```
I think that originally you had it as `HashSet`, now it's `List`. I assume you did it to preserve order. Is that correct? If that's the case then I don't think that `GroupByFields.Keys.ToList()` guarantees that. Note that we have `OrderedGroupByFields`

Anyway, I don't think that we should care about that on this level. If you need that somewhere then assure it there.

grisha-kotler replied:
Fixed. I'm looking for a consistent order (of any kind).
So if the group by keys in the function are something like `group by Name, LastName`, I want them in that order.

grisha-kotler replied:
`OrderedGroupByFields` doesn't preserve the original order.

grisha-kotler committed:
RavenDB-18164 - use the same order that we use for other indexes

ppekrol commented on src/Raven.Server/Documents/Indexes/MapReduce/Auto/AutoMapReduceIndexDefinition.cs:
```diff
@@ -19,13 +19,17 @@ internal class AutoMapReduceIndexDefinition : AutoIndexDefinitionBaseServerSide
 
             GroupByFields = groupByFields.ToDictionary(x => x.Name, x => x, StringComparer.Ordinal);
 
+            GroupByFieldNames = groupByFields.Select(x => x.Name).ToList();
```
I'm wondering if this is not safer to use here OrderedGroupByFields instead of groupByFields

arekpalinski commented:
Merging it. Grisha needs to check something with order of group by fields but current stat is good enough for perf testing.
